bpo-35196: optimize Squeezer's write() interception

taleinat opened the pull request:
This also cleans up and improves the testing of `count_lines()`, which has been significantly changed as part of this optimization.

<!-- issue-number: [bpo-35196](https://bugs.python.org/issue35196) -->
https://bugs.python.org/issue35196
<!-- /issue-number -->

taleinat committed:
bpo-35196: optimize Squeezer's write() interception

taleinat committed:
remove dead, old test code from test_squeezer.py

taleinat committed:
bpo-35196: fix test_reload() failing

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -231,25 +245,56 @@ def __init__(self, editwin):
+                # number of lines, but only if the given text is not too
+                # long for it to possibly succeed.
+                line_width = self.get_line_width()
+                if s_len < auto_squeeze_min_lines * line_width:
```
Having you done any timings to indicate that the extra work here pays off?

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -277,23 +322,26 @@ def count_lines(self, s):
         """
         # Tab width is configurable
```
As I said in the general comment, not true.

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -212,6 +215,17 @@ def reload(cls):
+        # rely on Tkinter's "default root", so each instance will load
+        # font info separately using its editor windows's Tk root.
+        new_instance_weakrefs = []
+        for inst_ref in cls._instance_weakrefs:
```
Since only the Squeezer attached to Shell is ever used (and that should be the only 1 created) I don't see the point of this new list and machinery.

taleinat replied:
For some reason I thought there was a way to have more than a single shell, but I was wrong.

We could make Squeezer a singleton and simplify this logic. I still think we'd need something of the sort, though, since the shell window can be closed and opened again.

I'll push an update soon.

taleinat replied:
See the latest change which simplifies `Squeezer.reload()`.

terryjreedy replied:
I want to sometime consider having a separate execution process for each file that is run, for the benefit of code completion.  That might or might not mean a separate Shell frame for each.  I would rather add code (as opposed to comments) needed for multiples shell only when needed, and would want to understand why.

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -231,25 +245,56 @@ def __init__(self, editwin):
+                # First, a very quick check to skip very short texts.
+                s_len = len(s)
+                auto_squeeze_min_lines = self.auto_squeeze_min_lines
+                if s_len < auto_squeeze_min_lines:
```
Good idea.  Many single prints are less that 50 chars.  I think we can do even better.  I believe that the minimum string that could trigger squeezing is 49 newlines + about (linelength//8 + 1) tabs or 49 newlines + (linelength + 1) other chars.  If we assume a minimum linelength of 40 and allow for tab, the magic number would be 55.  I am a bit inclined to use a higher cutoff of, say 80, and not worry if we do not a autosqueeze every eligible string with mostly empty lines.

taleinat replied:
My guiding principle here was to keep the details as clear and straightforward as possible. You're welcome to change that if you like. Note, however, that the number of lines to auto-squeeze is the only user-configurable parameter, and your suggestion would make its meaning vague.

As per optimization, I find that this PR makes the auto-squeezing feature practically instantaneous in all cases which would be covered by your suggestion, so I don't think any other change is necessary.

terryjreedy replied:
OK, without a convincing need to change this, leave it as is.

taleinat committed:
remove unnecessary additional check which doesn't improve performance

taleinat committed:
instantiate Squeezer only on PyShell windows

This is cleaner and allows simplifying Squeezer.reload().

taleinat committed:
hard-code tab width to 8 (always true for IDLE's shell)

terryjreedy committed:
Merge remote-tracking branch 'origin/master' into pr_10454

terryjreedy committed:
Add blurb and idlelib news.

terryjreedy committed:
Condense character width comment and update to 8.6.

terryjreedy committed:
In test, change deprecated assertEquals to assertEqual.

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -210,6 +214,14 @@ def reload(cls):
+        # on Tkinter's "default root", so the instance will reload
+        # font info using its editor windows's Tk root.
+        if cls._instance_weakref is not None:
+            instance = cls._instance_weakref()
```
When testing, _instance_weakref is currently always None, so this and the next 2 lines are skipped.

taleinat replied:
That was a bug! I've fixed the bug and improved `test_reload` for better coverage.

terryjreedy replied:
I am still puzzled about the need to add weakrefs and tests. You said something earlier about "since the shell window can be closed and opened again.".  But if a shell window is closed, the associated squeezer and its buttons should be deleted.  If not, I would rather fix the bug than add a workaround.  Creating a new PyShell instance will create an new Squeezer instance.  What have I missed?

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -210,6 +214,14 @@ def reload(cls):
 
+        # Loading the font info requires a Tk root. IDLE doesn't rely
+        # on Tkinter's "default root", so the instance will reload
+        # font info using its editor windows's Tk root.
```
This looks unneeded.  It seems to me that squeezer should just get the 'editor' font from the editor, as self.editwin.font or self.base_text.font, whichever, when it is needed.  When reloads are called by configdialog, the editors are prompted to update themselves also.  I don't see that squeezer should cache it separately.

taleinat replied:
The idea here is to avoid calling `Font.measure('0')` repeatedly, especially in the time-sensitive context of deciding whether to auto-squeeze. Since it can easily be computed and stored in advance, I created the new `load_font()` method, which is called by `Squeezer.__init__()` and `Squeezer.reload()`.

This could be simplified by increasing coupling: Adding logic to `PyShell` to call its squeezer's `load_font()` function after font updates, or just call its squeezer's `reload()` when appropriate.

taleinat replied:
Also, the `EditorWindow` and `Text` classes have no `font` attribute, and I haven't found something equivalent. `Font(text, name=text.cget('font'))` is the best I could find.

terryjreedy replied:
Yes, the font measure should be cached.  My concern was how to get it efficiently, though it does not matter so much in reload.  Sorry for the misdirection.  The text 'font' is an option rather than attribute, and it consists of a font tuple rather than Font instance.  I think the latter would actually be better for configuration, as well as Squeezer, but changing this would be another issue (like when mousewheel resizing is added).  I believe `Font(text, text['font']).measure('0') is what we need.  I may experiment.

terryjreedy committed:
Add coverage information.  Details in comment.

terryjreedy committed:
Add X windows comment.

terryjreedy committed:
Move PyShell only code to pyshell.PyShell.

terryjreedy committed:
Use existing local reference.

terryjreedy committed:
Fix comments (# Cap ... length ... end.) in squeezer.py.

terryjreedy committed:
Fix comments in test_squeezer to PEP standards.  Uncomment test
dependent on previous PR.  (It now passes.)

terryjreedy reviewed (changes requested):
Just a couple of things more.  For code, see inline comment.  For tests, see separate comment

taleinat committed:
fix bug in setting Squeezer._instance_weakref and improve test coverage

taleinat committed:
add a do-nothing test case for Squeezer.reload(), and fix test cleanup

taleinat committed:
re-format inline comments

terryjreedy committed:
Update test_squeezer, trigger retest

terryjreedy committed:
Reverse asssertions to test 2nd

csabella commented on Lib/idlelib/squeezer.py:
```diff
@@ -273,25 +300,24 @@ def count_lines(self, s):
+            self.base_text.winfo_width() - self.window_width_delta
+
+        # Divide the width of the Text widget by the font width,
+        # which is taked to be the width of '0' (zero).
```
s/taked/taken

terryjreedy replied:
fixed

terryjreedy committed:
Change taked to taken

taleinat committed:
fix test_reload() check that zero_with_char has increased

The check was much too restrictive, and failed on some platforms in
certain cases. Really, test_reload() just needs to check that the
reload() call had an effect.

terryjreedy commented on Lib/idlelib/squeezer.py:
```diff
@@ -273,25 +300,24 @@ def count_lines(self, s):
+    def load_font(self):
+        text = self.base_text
+        self.zero_char_width = \
+            Font(text, name=text.cget('font')).measure('0')
```
Bug.  The font name is a string added to tk's registry of fonts.  The tuple gets turned into a string.  Since I cannot accept a suggestion after making myself assignee (contrary to (i) information popup, I will change 'name' to 'font' in the web editor and see if it changes CI bots.  (Test still pass on my machine.)

terryjreedy committed:
Change 'name' to 'font' in load_font Font().

terryjreedy committed:
Reset font size by resetting font option.

taleinat committed:
minor code cleanup in Squeezer's mywrite()
